src/oidcop/oidc/token.py

```diff
@@ -352,6 +363,249 @@ def post_parse_request(
+            JWKESTException,
+        ) as err:
+            return self.endpoint.error_cls(
+                error="invalid_request", error_description="%s" % err
+            )
+
+        _mngr = _context.session_manager
+        try:
+            _session_info = _mngr.get_session_info_by_token(
+                request["subject_token"], grant=True
+            )
+        except (KeyError, UnknownToken):
+            logger.error("Subject token invalid.")
+            return self.error_cls(
+                error="invalid_request",
+                error_description="Subject token invalid"
+            )
+
+        token = _mngr.find_token(_session_info["session_id"], request["subject_token"])
+
+        resp = self.enforce_policy(request, token)
+        if isinstance(resp, TokenErrorResponse):
+            return resp
+
+        if token.is_active() is False:
+            return self.error_cls(
+                error="invalid_request", error_description="Subject token inactive"
+            )
+        return resp
+
+    def enforce_policy(self, request, token):
+        _context = self.endpoint.server_get("endpoint_context")
+
+        if not isinstance(token, (AccessToken, RefreshToken)):
+            return self.error_cls(
+                error="invalid_request", error_description="Wrong token type"
+            )
+
+        if request["subject_token_type"] not in self.config["token_exchange"]["subject_token_types_supported"]:
+            return TokenErrorResponse(
+                error="invalid_request",
+                error_description="Unsupported subject token type",
+            )
+
+        if (
+            "requested_token_type" in request
+            and request["requested_token_type"] not in self.config["token_exchange"]["requested_token_types_supported"]
+        ):
+            return TokenErrorResponse(
+                error="invalid_request",
+                error_description="Unsupported requested token type",
+            )
+
+        if (
+            "requested_token_type" in request
+            and request["requested_token_type"] == "urn:ietf:params:oauth:token-type:refresh_token"
+            and request["subject_token_type"] == "urn:ietf:params:oauth:token-type:access_token"
+            and "offline_access" not in token.scope
+        ):
+            return TokenErrorResponse(
+                error="invalid_request",
+                error_description="Exchange access token to refresh token forbbiden",
+            )
+
+        elif (
+            "requested_token_type" in request
+            and request["requested_token_type"] == "urn:ietf:params:oauth:token-type:refresh_token"
+            and request["subject_token_type"] != "urn:ietf:params:oauth:token-type:refresh_token"
+        ):
+            return TokenErrorResponse(
+                    error="invalid_request",
+                    error_description=f"Exchange {request['subject_token_type']} to refresh token forbbiden",
+                )
```

Comment:
Do we need to enforce this? Shouldn't this be configurable?

Reply:
I made this part of the default token exchange callable